core/lib/zksync_core/src/api_server/web3/tests/ws.rs

```diff
@@ -0,0 +1,694 @@
+    for _ in 0..expected_count {
+        let log = tokio::time::timeout(TEST_TIMEOUT, sub.next())
+            .await
+            .context("Timed out waiting for new log")?
+            .context("Logs subscription terminated")??;
+        logs.push(log);
+    }
+    Ok(logs)
+}
+
+fn assert_logs_match(actual_logs: &[api::Log], expected_logs: &[&VmEvent]) {
+    assert_eq!(actual_logs.len(), expected_logs.len());
+    for (actual_log, &expected_log) in actual_logs.iter().zip(expected_logs) {
+        assert_eq!(actual_log.address, expected_log.address);
+        assert_eq!(actual_log.topics, expected_log.indexed_topics);
+        assert_eq!(actual_log.data.0, expected_log.value);
+    }
+}
+
+#[tokio::test]
+async fn log_subscriptions() {
+    test_ws_server(LogSubscriptions).await;
+}
+
+#[derive(Debug)]
+struct LogSubscriptionsWithNewBlock;
+
+#[async_trait]
+impl WsTest for LogSubscriptionsWithNewBlock {
+    async fn test(&self, client: &WsClient, pool: &ConnectionPool) -> anyhow::Result<()> {
+        let Subscriptions {
+            mut all_logs_subscription,
+            mut address_subscription,
+            ..
+        } = Subscriptions::new(client).await?;
+        // Wait a little until subscriptions are fully registered.
+        tokio::time::sleep(POLL_INTERVAL).await;
+
+        let mut storage = pool.access_storage().await?;
+        let (_, events) = LogSubscriptions::update_storage(&mut storage, 1, 0).await?;
+        drop(storage);
+        let events: Vec<_> = events.iter().collect();
+
+        let all_logs = collect_logs(&mut all_logs_subscription, 4).await?;
+        assert_logs_match(&all_logs, &events);
+
+        // Create a new block and wait for the pub-sub notifier to run.
+        let mut storage = pool.access_storage().await?;
+        let (_, new_events) = LogSubscriptions::update_storage(&mut storage, 2, 4).await?;
+        drop(storage);
+        let new_events: Vec<_> = new_events.iter().collect();
+
+        let all_new_logs = collect_logs(&mut all_logs_subscription, 4).await?;
+        assert_logs_match(&all_new_logs, &new_events);
+
+        let address_logs = collect_logs(&mut address_subscription, 4).await?;
+        assert_logs_match(
+            &address_logs,
+            &[events[0], events[3], new_events[0], new_events[3]],
+        );
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn log_subscriptions_with_new_block() {
+    test_ws_server(LogSubscriptionsWithNewBlock).await;
+}
+
+#[derive(Debug)]
+struct LogSubscriptionsWithManyBlocks;
+
+#[async_trait]
+impl WsTest for LogSubscriptionsWithManyBlocks {
+    async fn test(&self, client: &WsClient, pool: &ConnectionPool) -> anyhow::Result<()> {
+        let Subscriptions {
+            mut all_logs_subscription,
+            mut address_subscription,
+            ..
+        } = Subscriptions::new(client).await?;
+        // Wait a little until subscriptions are fully registered.
+        tokio::time::sleep(POLL_INTERVAL).await;
```

Comment:
Sometimes in this code "Wait a little" waits for POLL_INTERVAL and sometimes double that; judging by the logic in `notify_*` are you absolutely sure that at least the former will not cause flakiness in the CI?

Reply:
I've removed timeouts in tests in favor of listening to events emitted by subscription logic. It required significant refactoring of that logic, but IMO, it improved it a bit 🙃